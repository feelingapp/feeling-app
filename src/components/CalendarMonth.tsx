import { CSSProperties } from "react"
import { styled } from "linaria/react"
import { addDays, format, getDaysInMonth, getISODay } from "date-fns/esm"

import CalendarDay from "./CalendarDay"
import { getDateFromIndex } from "../utils/dates"

const Container = styled.div`
  display: flex;
  justify-content: center;

  &:last-child {
    :after {
      content: "";
      display: block;
      position: absolute;
      right: 0;
      bottom: calc(-1 * var(--tab-bar-height));
      height: var(--tab-bar-height);
      width: 100%;
    }
  }
`

const Content = styled.div`
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  gap: clamp(4px, 100vw, 8px);
  margin-bottom: var(--space-m);
  justify-items: center;
  user-select: none;
  max-width: 600px;
  min-width: 270px;
  width: 100%;
`

const MonthTitle = styled.p`
  grid-column: -1 / 1;
  justify-self: center;
  margin-bottom: var(--space-xs);
  color: var(--grey-color);
  font-size: 1.6rem;
  line-height: 1;
`

const WeekDayTitle = styled.div`
  margin-bottom: var(--space-xxs);
  color: var(--grey-color);
  font-size: 1.2rem;
  text-align: center;
  line-height: 1;
`

interface Props {
  index: number
  style: CSSProperties
}

export default function Month(props: Props): JSX.Element {
  const date = getDateFromIndex(props.index)
  const numberOfDays = getDaysInMonth(date)
  const day = getISODay(date) % 7

  return (
    <Container style={props.style}>
      <Content>
        <MonthTitle>{format(date, "MMMM yyyy")}</MonthTitle>

        <WeekDayTitle>M</WeekDayTitle>
        <WeekDayTitle>T</WeekDayTitle>
        <WeekDayTitle>W</WeekDayTitle>
        <WeekDayTitle>T</WeekDayTitle>
        <WeekDayTitle>F</WeekDayTitle>
        <WeekDayTitle>S</WeekDayTitle>
        <WeekDayTitle>S</WeekDayTitle>

        {/* Add blank calendar days so actual ones are inline with week day titles */}
        {Array.from({ length: day }).map((_, index) => (
          <CalendarDay date={date} blank={true} key={index} />
        ))}

        {Array.from({ length: numberOfDays }).map((_, index) => (
          <CalendarDay date={addDays(date, index)} key={index} />
        ))}
      </Content>
    </Container>
  )
}
