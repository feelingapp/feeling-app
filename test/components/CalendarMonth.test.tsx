import { shallow } from "enzyme"

import CalendarMonth from "../../src/components/CalendarMonth"

describe("Calendar month: ", () => {
  it("matches snapshot", () => {
    const wrapper = shallow(
      <CalendarMonth index={1} style={{ width: 600, height: 200 }} />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
