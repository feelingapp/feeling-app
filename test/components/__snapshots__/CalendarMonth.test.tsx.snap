// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar month:  matches snapshot 1`] = `
<Container
  style={
    Object {
      "height": 200,
      "width": 600,
    }
  }
>
  <Content>
    <MonthTitle>
      February 2020
    </MonthTitle>
    <WeekDayTitle>
      M
    </WeekDayTitle>
    <WeekDayTitle>
      T
    </WeekDayTitle>
    <WeekDayTitle>
      W
    </WeekDayTitle>
    <WeekDayTitle>
      T
    </WeekDayTitle>
    <WeekDayTitle>
      F
    </WeekDayTitle>
    <WeekDayTitle>
      S
    </WeekDayTitle>
    <WeekDayTitle>
      S
    </WeekDayTitle>
    <CalendarDay
      blank={true}
      date={2020-02-01T00:00:00.000Z}
      key="0"
    />
    <CalendarDay
      blank={true}
      date={2020-02-01T00:00:00.000Z}
      key="1"
    />
    <CalendarDay
      blank={true}
      date={2020-02-01T00:00:00.000Z}
      key="2"
    />
    <CalendarDay
      blank={true}
      date={2020-02-01T00:00:00.000Z}
      key="3"
    />
    <CalendarDay
      blank={true}
      date={2020-02-01T00:00:00.000Z}
      key="4"
    />
    <CalendarDay
      blank={true}
      date={2020-02-01T00:00:00.000Z}
      key="5"
    />
    <CalendarDay
      date={2020-02-01T00:00:00.000Z}
      key="0"
    />
    <CalendarDay
      date={2020-02-02T00:00:00.000Z}
      key="1"
    />
    <CalendarDay
      date={2020-02-03T00:00:00.000Z}
      key="2"
    />
    <CalendarDay
      date={2020-02-04T00:00:00.000Z}
      key="3"
    />
    <CalendarDay
      date={2020-02-05T00:00:00.000Z}
      key="4"
    />
    <CalendarDay
      date={2020-02-06T00:00:00.000Z}
      key="5"
    />
    <CalendarDay
      date={2020-02-07T00:00:00.000Z}
      key="6"
    />
    <CalendarDay
      date={2020-02-08T00:00:00.000Z}
      key="7"
    />
    <CalendarDay
      date={2020-02-09T00:00:00.000Z}
      key="8"
    />
    <CalendarDay
      date={2020-02-10T00:00:00.000Z}
      key="9"
    />
    <CalendarDay
      date={2020-02-11T00:00:00.000Z}
      key="10"
    />
    <CalendarDay
      date={2020-02-12T00:00:00.000Z}
      key="11"
    />
    <CalendarDay
      date={2020-02-13T00:00:00.000Z}
      key="12"
    />
    <CalendarDay
      date={2020-02-14T00:00:00.000Z}
      key="13"
    />
    <CalendarDay
      date={2020-02-15T00:00:00.000Z}
      key="14"
    />
    <CalendarDay
      date={2020-02-16T00:00:00.000Z}
      key="15"
    />
    <CalendarDay
      date={2020-02-17T00:00:00.000Z}
      key="16"
    />
    <CalendarDay
      date={2020-02-18T00:00:00.000Z}
      key="17"
    />
    <CalendarDay
      date={2020-02-19T00:00:00.000Z}
      key="18"
    />
    <CalendarDay
      date={2020-02-20T00:00:00.000Z}
      key="19"
    />
    <CalendarDay
      date={2020-02-21T00:00:00.000Z}
      key="20"
    />
    <CalendarDay
      date={2020-02-22T00:00:00.000Z}
      key="21"
    />
    <CalendarDay
      date={2020-02-23T00:00:00.000Z}
      key="22"
    />
    <CalendarDay
      date={2020-02-24T00:00:00.000Z}
      key="23"
    />
    <CalendarDay
      date={2020-02-25T00:00:00.000Z}
      key="24"
    />
    <CalendarDay
      date={2020-02-26T00:00:00.000Z}
      key="25"
    />
    <CalendarDay
      date={2020-02-27T00:00:00.000Z}
      key="26"
    />
    <CalendarDay
      date={2020-02-28T00:00:00.000Z}
      key="27"
    />
    <CalendarDay
      date={2020-02-29T00:00:00.000Z}
      key="28"
    />
  </Content>
</Container>
`;
